# Student Gradebook Analyzer

Workshop project 2 - AIML using Python

This notebook analyzes a class gradebook using Pandas and NumPy. I have 10 students and their scores across 4 assignments. The plan is:
1. Load the data
2. Get stats for each student (avg/median/max/min)
3. Get stats for each assignment (avg/median/max/min)
4. Sort students to find top performers and who needs help
5. Use NumPy for standard deviation to see how spread out the scores are
6. Save everything to a new CSV

In [24]:
import pandas as pd
import numpy as np

## 1. Loading the data

I made a small CSV file with 10 students and 4 assignments (out of 100 each). Honestly just typed in random-ish realistic looking scores for this, this isn't real data.

In [12]:
df = pd.read_csv("gradebook.csv")
df

,Student_Name,Assignment_1,Assignment_2,Assignment_3,Assignment_4
0,Aarav Sharma,78,85,90,88
1,Priya Mehta,92,89,95,91
2,Rohan Verma,65,70,68,72
3,Sneha Iyer,88,91,85,93
4,Karan Singh,55,60,58,62
5,Ananya Gupta,95,97,93,96
6,Vikram Rao,72,68,75,70
7,Isha Kapoor,80,82,79,84
8,Aditya Joshi,60,65,63,59
9,Tanya Nair,85,88,90,87


## 2. Per-student statistics

For each student, I need the average, median, max and min across their 4 assignments. The assignment columns are side by side (columns), so to get a stat *per row* (per student) I need `axis=1`. This confused me at first ngl, kept mixing up axis=0 and axis=1, but the way I remember it now is: axis=1 squishes the columns together for each row.

In [13]:
assignment_cols = ["Assignment_1", "Assignment_2", "Assignment_3", "Assignment_4"]

df["Average"] = df[assignment_cols].mean(axis=1)
df["Median"] = df[assignment_cols].median(axis=1)
df["Max_Score"] = df[assignment_cols].max(axis=1)
df["Min_Score"] = df[assignment_cols].min(axis=1)

df

,Student_Name,Assignment_1,Assignment_2,Assignment_3,Assignment_4,Average,Median,Max_Score,Min_Score
0,Aarav Sharma,78,85,90,88,85.25,86.5,90,78
1,Priya Mehta,92,89,95,91,91.75,91.5,95,89
2,Rohan Verma,65,70,68,72,68.75,69.0,72,65
3,Sneha Iyer,88,91,85,93,89.25,89.5,93,85
4,Karan Singh,55,60,58,62,58.75,59.0,62,55
5,Ananya Gupta,95,97,93,96,95.25,95.5,97,93
6,Vikram Rao,72,68,75,70,71.25,71.0,75,68
7,Isha Kapoor,80,82,79,84,81.25,81.0,84,79
8,Aditya Joshi,60,65,63,59,61.75,61.5,65,59
9,Tanya Nair,85,88,90,87,87.50,87.5,90,85


## 3. Per-assignment statistics

Now flip it - for each assignment, what's the average/median/max/min across all 10 students? This time it's column-wise so I don't even need to pass axis (default is axis=0).

In [14]:
assignment_stats = df[assignment_cols].agg(["mean", "median", "max", "min"])
assignment_stats

,Assignment_1,Assignment_2,Assignment_3,Assignment_4
mean,77.0,79.5,79.6,80.2
median,79.0,83.5,82.0,85.5
max,95.0,97.0,95.0,96.0
min,55.0,60.0,58.0,59.0


Assignment_4 has the highest average (80.2) and Assignment_1 has the lowest (77.0). Makes sense if people got more comfortable as the assignments went on, or maybe Assignment_1 was just harder.

## 4. Sorting students by average score

In [15]:
top_students = df.sort_values(by="Average", ascending=False)
top_students[["Student_Name", "Average"]]

,Student_Name,Average
5,Ananya Gupta,95.25
1,Priya Mehta,91.75
3,Sneha Iyer,89.25
9,Tanya Nair,87.50
0,Aarav Sharma,85.25
7,Isha Kapoor,81.25
6,Vikram Rao,71.25
2,Rohan Verma,68.75
8,Aditya Joshi,61.75
4,Karan Singh,58.75


### Top 3 performers

In [16]:
top_3 = top_students.head(3)
top_3[["Student_Name", "Average"]]

,Student_Name,Average
5,Ananya Gupta,95.25
1,Priya Mehta,91.75
3,Sneha Iyer,89.25


### Bottom 3 (might need extra support)

Same thing, sorted the other way. Could be useful if a teacher wanted to know who to check in with.

In [17]:
bottom_3 = df.sort_values(by="Average", ascending=True).head(3)
bottom_3[["Student_Name", "Average"]]

,Student_Name,Average
4,Karan Singh,58.75
8,Aditya Joshi,61.75
2,Rohan Verma,68.75


## 5. Using NumPy properly

I imported numpy earlier but wasn't actually using it for anything (pandas was doing all the work). So here I'm pulling the scores out into an actual NumPy array with `.to_numpy()` and calculating standard deviation - this tells us how spread out the scores are. A high std means some students scored really differently from others on that assignment, a low std means everyone scored pretty similarly.

In [18]:
scores_array = df[assignment_cols].to_numpy()

class_average_np = np.mean(scores_array)
class_std_np = np.std(scores_array)

print(f"Overall class average (via NumPy): {class_average_np:.2f}")
print(f"Overall class standard deviation (via NumPy): {class_std_np:.2f}")

Overall class average (via NumPy): 79.08
Overall class standard deviation (via NumPy): 12.61


In [19]:
# std per assignment, axis=0 because I want one number per column this time
assignment_std = np.std(scores_array, axis=0)

for col_name, std_val in zip(assignment_cols, assignment_std):
    print(f"{col_name}: {std_val:.2f}")

Assignment_1: 12.98
Assignment_2: 12.04
Assignment_3: 12.49
Assignment_4: 12.66


## 6. Saving the results

In [22]:
df.to_csv("gradebook_with_stats.csv", index=False)
print("Saved to gradebook_with_stats.csv")

Saved to gradebook_with_stats.csv


## Summary

- Ananya Gupta has the highest average (95.25), followed by Priya Mehta (91.75) and Sneha Iyer (89.25)
- Karan Singh has the lowest average (58.75) and might need extra support, along with Aditya Joshi and Rohan Verma
- Assignment_4 had the highest class average (80.2), Assignment_1 had the lowest (77.0)
- Overall class average across everything comes out to 79.08
- Assignment_1 has the highest standard deviation (12.98), meaning scores were the most spread out / inconsistent on that one

### What I actually learned doing this
Biggest thing was getting `axis=0` vs `axis=1` straight in my head for Pandas aggregations - row-wise needs axis=1, column-wise is the default (axis=0). Also learned `.agg()` lets you run multiple stat functions in one line instead of calling `.mean()`, `.median()` etc separately every time, which is way less repetitive. And finally figured out a real use for NumPy instead of just importing it because the project said to - using `.to_numpy()` to convert a DataFrame slice into an array, then `np.std()` to measure how spread out scores were.